mqtt reconnect improvements

kanstrup opened the pull request.

david-cermak commented on mqtt_client.c:
```diff
@@ -744,8 +745,11 @@ static void esp_mqtt_task(void *pv)
+                    break;
                 }
-                vTaskDelay(client->wait_timeout_ms / 2 / portTICK_RATE_MS);
+                xEventGroupClearBits(client->status_bits, RECONNECT_BIT);
```
What is the purpose of clearing bits here? Wouldn't we miss reconnect command if it executed other than while waiting in the below EventWaitBits?

I know that the `esp-mqtt` is not yet thread safe, but what about typical call of `esp_mqtt_client_reconnect` on disconnect event? My suspicion is that it would be missed. Would not it make more sense to put it under `MQTT_STATE_INIT` ?

kanstrup replied:
Indeed clearing RECONNECT_BIT here looks fishy. I've been running with this patch applied locally for quite some time so can't remember why I decided to clear it here. I think I just needed the bit to be cleared somewhere for subsequent reconnects to function properly.

Agree that it makes more sense to put it under MQTT_STATE_INIT instead. Will upload an updated version shortly.

david-cermak commented:
Hi @kanstrup 

Thanks for your contribution. I left one comment, but otherwise it looks good!

kanstrup commented:
Hi @david-cermak, I updated the code based on your comment. Also rebased on top of idf branch.

It appears esp-idf is not using the top of idf branch but rather a sha1 a few commits earlier. Travis report problems both if I base on top of idf branch and base it on the sha1 used by esp-idf.

If based on sha1 esp-idf uses travis report a merge conflict with changes on idf branch.
If based on idf branch the mqtt example code does not build due to recently added MQTT_EVENT_BEFORE_CONNECT is not handled in esp-idf mqtt example code.

Fixing the example code is simple. Something like this:

+        case MQTT_EVENT_BEFORE_CONNECT:
+            ESP_LOGI(TAG, "MQTT_EVENT_BEFORE_CONNECTED");
+            break; 

Please let me know if there's anything I can do to help the situation.

david-cermak commented:
@kanstrup Thanks for the update. LGTM and could be merged. Still one more rebase might be needed, as there are 4 commits pending in Espressif internal repository.

Yeah, the travis is broken, now. The trouble is that the mqtt examples are in `esp-idf` and a fix similar to what you're suggesting is pending in idf internal repo. Sorry for the inconvenience, it was introduced while not having CI functional internally.

kanstrup commented:
> can you please rebase your work onto idf branch? Travis should be fixed and if it passes the CI we can merge it.

@david-cermak Just rebased but CI still fails. Can you have a look at it? Seems not related to my changes.

david-cermak commented:
@kanstrup I am sorry. the issue is related to fetching the changes from PRs (works properly for esp-mqtt repo, but not for forks), this needs a fix from our side. will deliver in few days.
I know that it takes so long, please hold on...

david-cermak commented:
@kanstrup could you please rebase again? CI should pass now.

tuanpmt commented on mqtt_client.c:
```diff
@@ -878,6 +882,18 @@ esp_err_t esp_mqtt_client_start(esp_mqtt_client_handle_t client)
+{
+    ESP_LOGI(TAG, "Client force reconnect requested");
+    if (client->state != MQTT_STATE_WAIT_TIMEOUT) {
+        ESP_LOGD(TAG, "Client not waiting for reconnect. Ignore request");
```
```suggestion
        ESP_LOGD(TAG, "The client is not waiting for reconnecting. Ignore the request");
```

mikaelkanstrup replied:
Ok, updated.

tuanpmt commented on mqtt_client.c:
```diff
@@ -878,6 +882,18 @@ esp_err_t esp_mqtt_client_start(esp_mqtt_client_handle_t client)
+    ESP_LOGI(TAG, "Client force reconnect requested");
+    if (client->state != MQTT_STATE_WAIT_TIMEOUT) {
+        ESP_LOGD(TAG, "Client not waiting for reconnect. Ignore request");
+        return ESP_OK;
```
I think it may return `ESP_FAIL` useful for the application needs to know the execution results

mikaelkanstrup replied:
Agree. Done.
